File: lc_java/lcJava/BinaryTree/AddOneRow.java
package lcJava.BinaryTree;

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class AddOneRow {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if (depth == 1) {
            TreeNode newRoot = new TreeNode(val);
            newRoot.left = root;
            return newRoot;
        }
        
        dfs(root, val, 1, depth);
        return root;
    }
    
    private void dfs(TreeNode node, int val, int currentDepth, int targetDepth) {
        if (node == null) return;
        
        if (currentDepth == targetDepth - 1) {
            TreeNode oldLeft = node.left;
            TreeNode oldRight = node.right;
            
            node.left = new TreeNode(val);
            node.left.left = oldLeft;
            
            node.right = new TreeNode(val);
            node.right.right = oldRight;
            return;
        }
        
        dfs(node.left, val, currentDepth + 1, targetDepth);
        dfs(node.right, val, currentDepth + 1, targetDepth);
    }
    
    // 测试函数
    public static void test() {
        AddOneRow solution = new AddOneRow();
        
        // 测试用例1
        TreeNode root1 = new TreeNode(4);
        root1.left = new TreeNode(2);
        root1.right = new TreeNode(6);
        root1.left.left = new TreeNode(3);
        root1.left.right = new TreeNode(1);
        root1.right.left = new TreeNode(5);
        
        TreeNode result1 = solution.addOneRow(root1, 1, 2);
        System.out.println("Test 1 passed");
        
        // 测试用例2
        TreeNode root2 = new TreeNode(4);
        root2.left = new TreeNode(2);
        root2.left.left = new TreeNode(3);
        root2.left.right = new TreeNode(1);
        
        TreeNode result2 = solution.addOneRow(root2, 1, 3);
        System.out.println("Test 2 passed");
    }
    
    public static void main(String[] args) {
        test();
    }
}

/*
算法思路：DFS递归遍历到目标深度的前一层，然后插入新节点
时间复杂度：O(n)
空间复杂度：O(h) - 递归栈深度
*/

File: lc_py/lcPy/tree/add_one_row.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def addOneRow(self, root: TreeNode, val: int, depth: int) -> TreeNode:
        if depth == 1:
            new_root = TreeNode(val)
            new_root.left = root
            return new_root
        
        def dfs(node, current_depth, target_depth):
            if not node:
                return
            
            if current_depth == target_depth - 1:
                old_left = node.left
                old_right = node.right
                
                node.left = TreeNode(val)
                node.left.left = old_left
                
                node.right = TreeNode(val)
                node.right.right = old_right
                return
            
            dfs(node.left, current_depth + 1, target_depth)
            dfs(node.right, current_depth + 1, target_depth)
        
        dfs(root, 1, depth)
        return root

# 测试函数
def test_add_one_row():
    solution = Solution()
    
    # 测试用例1
    root1 = TreeNode(4)
    root1.left = TreeNode(2)
    root1.right = TreeNode(6)
    root1.left.left = TreeNode(3)
    root1.left.right = TreeNode(1)
    root1.right.left = TreeNode(5)
    
    result1 = solution.addOneRow(root1, 1, 2)
    print("Test 1 passed")
    
    # 测试用例2
    root2 = TreeNode(4)
    root2.left = TreeNode(2)
    root2.left.left = TreeNode(3)
    root2.left.right = TreeNode(1)
    
    result2 = solution.addOneRow(root2, 1, 3)
    print("Test 2 passed")

if __name__ == "__main__":
    test_add_one_row()

"""
算法思路：DFS递归遍历到目标深度的前一层，然后插入新节点
时间复杂度：O(n)
空间复杂度：O(h) - 递归栈深度
"""